### PR TITLE
Make sure that Postgres always get a consistent snapshot

### PR DESCRIPTION
**Pull request by judahrand**
## Problem

This is a reworking of #24. 

## Proposed changes

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 
If it fixes a bug or resolves a feature request, be sure to link to that issue._

## Types of changes

What types of changes does your code introduce to PipelineWise?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] Description above provides context of the change
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Unit tests for changes (not needed for documentation changes)
- [ ] CI checks pass with my changes
- [ ] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [ ] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [ ] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions

### CODE REVIEW

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 1]
```diff
@@ -317,9 +323,7 @@ def fetch_current_log_pos(self):
                     'Logical replication not supported before PostgreSQL 9.4'
                 )
 
-        current_lsn = result[0].get('current_lsn')
-        file, index = current_lsn.split('/')
-        lsn = (int(file, 16) << 32) + int(index, 16)
+        lsn = parse_lsn(result[0].get('current_lsn'))
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py]: Is this a `dict.get`? If so, should `parse_lsn` accept `Optional[str]`?

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 2]
```diff
@@ -287,6 +288,33 @@ def get_current_lsn(self) -> int:
...
+        )
+
+        res = self.primary_host_query(
+            textwrap.dedent(
+                f"""SELECT *
+                    FROM pg_replication_slots
+                    WHERE slot_name = '{slot_name}'
+                    AND plugin = 'wal2json'
+                    AND slot_type = 'logical'"""
+            )
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py]: I don't think dedent is going to dedent this:
```python
In [1]: import textwrap

In [2]: textwrap.dedent("Foo\n    Bar")
Out[2]: 'Foo\n    Bar'

In [3]: textwrap.dedent("    Foo\n    Bar")
Out[3]: 'Foo\nBar'
```

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 3]
```diff
@@ -348,18 +349,33 @@ def fetch_current_log_pos(self):
...
+            # the oldest LSN which is deliverable by the replication slot.
+            time_waited = 0  # seconds
+            wait_period = 1  # seconds
+            max_time_waited = 60  # seconds
+            while oldest_lsn > current_lsn:
+                time.sleep(wait_period)
+                time_waited += wait_period
+                if time_waited > max_time_waited:
+                    raise RuntimeError('Replica database is lagging too far behind Primary.')
+                current_lsn = self.get_current_lsn()
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py]: Just to check -- can the oldest LSN change while this is happening? Or is that pinned once the connection is started?

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 4]
```diff
@@ -348,18 +349,33 @@ def fetch_current_log_pos(self):
...
+        current_lsn = self.get_current_lsn()
 
-        return {'lsn': lsn, 'version': 1}
+        if self.connection_config.get('replica_host'):
+            # Ensure that the newest LSN availiable on the replica is newer than
+            # the oldest LSN which is deliverable by the replication slot.
+            time_waited = 0  # seconds
+            wait_period = 1  # seconds
+            max_time_waited = 60  # seconds
+            while oldest_lsn > current_lsn:
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:367]: Are there any wrap-around issues we need to worry about here?

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 5]
```diff
@@ -348,18 +349,33 @@ def fetch_current_log_pos(self):
...
+                time.sleep(wait_period)
+                time_waited += wait_period
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py]: Strictly speaking this should probably use `time.time` to check the time passed, though I'm guessing the goal here is mostly to have a dumb timeout? If that's the case I wonder if a `for _ in range(...)` loop might be clearer (though admittedly slightly harder to configure)?

**Review by PeterJCLaw — APPROVED** [review]
Looks good generally. Just the question about wrap-around which I think needs addressing.

Just to check I understand the changes -- the main change here is the checking that the LSN on offer from the primary is suitable given the data that can be fetched from the replica. Is that right?

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 1 above)
Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py]: Yup, was a copy-pasta from elsewhere. Shouldn't be a `get`.

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 4 above)
Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py:367]: Postgres LSN is strictly monotonically increasing: https://www.postgresql.org/docs/11/wal-internals.html

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 3 above)
Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py]: `oldest_lsn` is really `oldest_lsn_which_is_required_for_consistent_replication` which is either provided by Postgres when the replication slot is created or required as the most recent `confirmed_flush_lsn` for the slot. `confirmed_flush_lsn` cannot change while this is happening as PipelineWise is the only application which should be talking to this slot and it is busy doing this.

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 2 above)
Comment (PeterJCLaw) [reply pipelinewise/fastsync/commons/tap_postgres.py]: https://github.com/thread/pipelinewise/pull/25/commits/783e28ee9068ea6516f65a1895dc2d22e3830a29 appears to address this.

Note however that `textwrap.dedent` _can_ handle leading non-indented newlines:
``` python
In [6]: textwrap.dedent("\n    Foo\n    Bar")
Out[6]: '\nFoo\nBar'
```
So the leading `\` in the current version of this is not needed.

Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py]: You end up with an extra `\n` at the start then though. The leading `\` makes the leading `\n` go away.

**Pull request merged**